Enabling Transformer fast path for not batch_first

soumyodipto opened the pull request:
Summary: The fast path for the `forward()` method in `MultiheadAttention` only accepted `batch_first = True`. This diff enables fast path for `batch_first=False` as well.

Test Plan:
Added unit test for fast path for both values of `batch_first` producing identical outputs.

`buck test mode/dev-nosan //caffe2/torch/nn/modules/tests:test_activation` passed

Reviewed By: mikekgfb

Differential Revision: D39669982

zrphercule reviewed (approved):
Thanks for this!

albanD commented on torch/nn/modules/tests/test_activation.py:
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python3
```
Could you move this next to the other tests in `test/` please?

soumyodipto replied:
See @cpuhrsch 's comment below. I'm moving it to /test/test_native_mha.py since this change enables a new run mode for native mha.

albanD commented on torch/nn/modules/activation.py:
```diff
@@ -1168,10 +1190,10 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
-            return attn_output.transpose(1, 0), attn_output_weights
-        else:
-            return attn_output, attn_output_weights
+            if self.batch_first and is_batched:
```
Why duplicate this code in both branches?

albanD commented on torch/nn/modules/activation.py:
```diff
@@ -1117,21 +1115,41 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
+                        average_attn_weights,
+                        1 if key_padding_mask is not None else 0 if attn_mask is not None else None)
+                else:  # If batch_first is not true we need to transpose the input
+                    # for fast-path non self-attention is used (query, key value are same tensors)
```
nit: query, key and value

soumyodipto replied:
Thanks, made the change

albanD commented on torch/nn/modules/activation.py:
```diff
@@ -1117,21 +1115,41 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
+                        1 if key_padding_mask is not None else 0 if attn_mask is not None else None)
+                else:  # If batch_first is not true we need to transpose the input
+                    # for fast-path non self-attention is used (query, key value are same tensors)
+                    query = key = value = query.transpose(1, 0)
```
You can have this line and the one that transpose back in a `if not self.batch_first` so that you don't have to duplicate the rather verbose call to `_native_multi_head_attention`.

soumyodipto replied:
The reason why I chose to duplicate the verbose call to `_native_multi_head_attention` is for performance reasons since it would do only one check for `self.batch_first` instead of two. Since performance is key, I chose this approach over code conciseness.

cpuhrsch replied:
Did you confirm this makes a difference in performance? I highly, highly doubt it. `self.batch_first` is a bool.

soumyodipto replied:
No it did not make a difference.

cpuhrsch commented on test/test_activation.py:
```diff
@@ -0,0 +1,30 @@
+import unittest
+from torch.nn.modules.activation import MultiheadAttention
+
+class TestActivation(unittest.TestCase):
```
This should fit neatly into one of our existing test files

soumyodipto replied:
I'm moving it to `/test/test_native_mha.py`

soumyodipto committed:
Enabling Transformer fast path for not batch_first (#85576)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/85576

The fast path for the `forward()` method in `MultiheadAttention` only accepted `batch_first = True`. This diff enables fast path for `batch_first=False` as well.

Modified `attention_test.py` in glow to resolve failure due to diff. See D40237969

Test Plan: Added unit test for fast path for both values of `batch_first` producing identical outputs. Test location - `//caffe2/test/test_native_mha.py`

Differential Revision: D39669982

fbshipit-source-id: fdb999941bc4732bd1bcd103906db576cf4cf27a

soumyodipto commented:
This is the updated diff with @cpuhrsch and @albanD's suggested changes
